### PR TITLE
Reported content

### LINKED ISSUE #3252: reported content: send notifications to admins when someone reports content (Trac #3252)
_Original ticket http://trac.elgg.org/ticket/3252 on 41236133-11-25 by cash, assigned to unknown._

Elgg version: Github Master

The reported contented plugin is used to report users who are spamming, trolling, or just being difficult. It is also used by users to report problems with a site. Currently, the only way an administrator knows about the reports is to login to the admin area and check.

Instead, when a report is created a notification should be sent to the admin (or admins) of the site. There should also be a plugin setting for turning this off.

### PR DESCRIPTION
**Pull request by jeabakker**
fixes #3252

### CODE REVIEW

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	/**
+	 * {@inheritDoc}
+	 */
+	public function getSubscriptions(): array {
+		$admins = elgg_get_entities([
+			'type' => 'user',
+			'limit' => false,
+			'batch' => true,
+			'metadata_name_value_pairs' => [
+				[
```
Comment (jdalsem) [inline mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:19]: you know you can write these pairs more readable?
`'admin' => 'yes'`

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+					'name' => 'admin_notification',
+					'value' => 1,
+				],
+			],
+			'private_setting_name_prefix' => 'plugin:user_setting:reportedcontent:',
+			'wheres' => [
+				function (QueryBuilder $qb, $main_alias) {
+					// no idea why an admin would report content and not fix it themselfs, but just in case
+					return $qb->compare("{$main_alias}.guid", '!=', $this->event->getActorGUID(), ELGG_VALUE_GUID);
```
Comment (jdalsem) [inline mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:34]: maybe this check is 'cheaper' in the next foreach loop??

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+		}
+		
+		return $subscriptions;
+	}
+	
+	/**
+	 * {@inheritDoc}
+	 */
+	public static function isConfigurableByUser(): bool {
+		return false;
```
Comment (jdalsem) [inline mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:51]: why not have this as a notification preference again (so not a configurable event, but as a 'default' preference)

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	
+	/**
+	 * {@inheritDoc}
+	 */
+	protected function getNotificationSubject(\ElggUser $recipient, string $method): string {
+		return elgg_echo('reportedcontent:notifications:create:admin:subject', [], $recipient->getLanguage());
```
Comment (jdalsem) [inline mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:58]: is this needed? wasn't there magic?

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	
+	protected function getNotificationSummary(\ElggUser $recipient, string $method): string {
+		return elgg_echo('reportedcontent:notifications:create:admin:summary', [], $recipient->getLanguage());
```
Comment (jdalsem) [inline mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:62]: is this needed? wasn't there magic?

File: mod/reportedcontent/languages/en.php [hunk 6]
```diff
@@ -23,10 +24,27 @@
 	'reportedcontent:report' => 'Report this',
 	'reportedcontent:numbertodisplay' => 'Number of reports to display',
 	'reportedcontent:user' => 'Report user',
+	'reportedcontent:newest' => 'Newest reports',
```
Comment (jdalsem) [inline mod/reportedcontent/languages/en.php:27]: New reports

File: mod/reportedcontent/languages/en.php [hunk 7]
```diff
@@ -23,10 +24,27 @@
...
+	'reportedcontent:archived' => 'Archived reports',
 
 	'reportedcontent:failed' => 'Sorry, the attempt to report this content has failed.',
 	'reportedcontent:notarchived' => 'We were not able to archive that report',
 	
 	'widgets:reportedcontent:name' => 'Reported content',
 	'widgets:reportedcontent:description' => 'Display reported content',
+	
+	'reportedcontent:usersettings:admin_notification' => 'Receive a notifcation when content is reported',
```
Comment (jdalsem) [inline mod/reportedcontent/languages/en.php:36]: notification typo

File: mod/reportedcontent/languages/en.php [hunk 8]
```diff
@@ -23,10 +24,27 @@
...
+	'reportedcontent:usersettings:admin_notification:help' => 'When a user reports content get an email notification',
```
Comment (jdalsem) [inline mod/reportedcontent/languages/en.php:37]: Why help text if it is same as title

File: mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php [hunk 9]
```diff
@@ -1,15 +1,29 @@
 <?php
 /**
- * Elgg Reported content admin page
+ * Elgg Reported content admin page (newest reports)
```
Comment (jdalsem) [inline mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php:3]: new*

File: mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php [hunk 10]
```diff
@@ -1,15 +1,29 @@
...
  */
 
+echo elgg_view('navigation/tabs', [
+	'tabs' => [
+		[
+			'text' => elgg_echo('reportedcontent:newest'),
```
Comment (jdalsem) [inline mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php:9]: reportedcontent:newest -> reportedcontent:new

File: mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php [hunk 11]
```diff
@@ -1,15 +1,29 @@
...
+
 echo elgg_list_entities([
 	'type' => 'object',
 	'subtype' => 'reported_content',
-	'order_by_metadata' => [
-		'name' => 'state',
-		'direction' => 'ASC',
-		'as' => 'text',
+	'metadata_name_value_pairs' => [
+		[
```
Comment (jdalsem) [inline mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php:23]: 'state' => 'active'

File: mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent/archive.php [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+		],
+	],
+]);
+
+echo elgg_list_entities([
+	'type' => 'object',
+	'subtype' => 'reported_content',
+	'metadata_name_value_pairs' => [
+		[
+			'name' => 'state',
```
Comment (jdalsem) [inline mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent/archive.php:24]: 'state' => 'archived'

File: mod/reportedcontent/views/default/widgets/reportedcontent/content.php [hunk 13]
```diff
@@ -6,17 +6,16 @@
...
 	'type' => 'object',
 	'subtype' => 'reported_content',
 	'limit' => (int) $widget->num_display ?: 4,
 	'pagination' => false,
-	'order_by_metadata' => [
-		'name' => 'state',
-		'direction' => 'ASC',
-		'as' => 'text',
+	'metadata_name_value_pairs' => [
+		[
```
Comment (jdalsem) [inline mod/reportedcontent/views/default/widgets/reportedcontent/content.php:15]: 'state' => 'active'

File: views/default/forms/plugins/settings/save.php [hunk 14]
```diff
@@ -9,13 +9,17 @@
...
-if (elgg_view_exists("plugins/$plugin_id/{$type}settings")) {
-	echo elgg_view("plugins/$plugin_id/{$type}settings", $vars);
+if (elgg_view_exists("plugins/{$plugin_id}/{$type}settings")) {
+	$settings = elgg_view("plugins/{$plugin_id}/{$type}settings", $vars);
+	if (empty($settings)) {
+		return;
+	}
+	
+	echo $settings;
 }
```
Comment (jdalsem) [inline views/default/forms/plugins/settings/save.php:23]: why not return if view does not exists?

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php (hunk 1 above)
Comment (jeabakker) [reply mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:19]: not going to change this

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php (hunk 2 above)
Comment (jeabakker) [reply mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:34]: I think not

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php (hunk 3 above)
Comment (jeabakker) [reply mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:51]: This only applies to detailed notification setting. This is not something which can be configured on an entity level

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php (hunk 4 above)
Comment (jeabakker) [reply mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:58]: yes, but gets more params (yes you could ignore those. this is more explicit

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php (hunk 5 above)
Comment (jeabakker) [reply mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:62]: not for summary, it defaults to empty string

File: mod/reportedcontent/languages/en.php (hunk 6 above)
Comment (jeabakker) [reply mod/reportedcontent/languages/en.php:27]: changed

File: mod/reportedcontent/languages/en.php (hunk 7 above)
Comment (jeabakker) [reply mod/reportedcontent/languages/en.php:36]: changed

File: mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php (hunk 9 above)
Comment (jeabakker) [reply mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php:3]: changed

File: mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php (hunk 10 above)
Comment (jeabakker) [reply mod/reportedcontent/views/default/admin/administer_utilities/reportedcontent.php:9]: changed

File: views/default/forms/plugins/settings/save.php (hunk 14 above)
Comment (jeabakker) [reply views/default/forms/plugins/settings/save.php:23]: changed

File: mod/reportedcontent/languages/en.php (hunk 8 above)
Comment (jeabakker) [reply mod/reportedcontent/languages/en.php]: removed

File: mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php (hunk 3 above)
Comment (jeabakker) [reply mod/reportedcontent/classes/Elgg/ReportedContent/Notifications/CreateReportedContentNotificationHandler.php:56]: changed it to a generic notification setting

**Review by jdalsem — APPROVED** [review]

**Pull request merged**